src/core_ocean/shared/mpas_ocn_equation_of_state_wright.F

```diff
@@ -0,0 +1,717 @@
+!>  that approximates the nonlinear equation of state using a set of
+!>  rational polynomials.
+!>  See Wright (1997), doi: 10.1175/1520-0426(1997)014<0735:AEOSFU>2.0.CO;2
+!
+!-----------------------------------------------------------------------
+
+module ocn_equation_of_state_wright
+
+   use mpas_kind_types
+   use mpas_derived_types
+   use mpas_pool_routines
+   use mpas_dmpar
+   use mpas_log
+   use mpas_constants
+   use ocn_constants
+   use ocn_config
+
+   implicit none
+   private
+   save
+
+   !--------------------------------------------------------------------
+   !
+   ! Public parameters
+   !
+   !--------------------------------------------------------------------
+
+   !--------------------------------------------------------------------
+   !
+   ! Public member functions
+   !
+   !--------------------------------------------------------------------
+
+   public :: ocn_equation_of_state_wright_density, &
+             ocn_equation_of_state_wright_init
+
+   !*** generic interface for case of density only or density and
+   !*** expansion coeffs
+   interface ocn_equation_of_state_wright_density
+      module procedure ocn_equation_of_state_wright_density_only
+      module procedure ocn_equation_of_state_wright_density_exp
+   end interface
+
+   !--------------------------------------------------------------------
+   !
+   ! Private module variables
+   !
+   !--------------------------------------------------------------------
+
+   !*** temporary array to hold modified T, S, p for EOS calculation
+   !*** declared here to make it thread-shared - could be replaced
+   !***   by a local subroutine temporary in a different thread model
+
+   real (kind=RKIND), dimension(:,:), allocatable :: &
+      tracerTemp, tracerSalt, boussinesqPres
+
+   !*** valid range of T,S for Wright (1997) EOS
+   !***   allowing T to become more negative because of sub-ice-shelf cavities
+
+   real (kind=RKIND), parameter :: &
+      ocnEqStateTmin = -3.0_RKIND, &! valid pot. temp. range
+      ocnEqStateTmax = 30.0_RKIND, &
+      ocnEqStateSmin = 28.0_RKIND, &! valid salinity, in psu
+      ocnEqStateSmax = 38.0_RKIND
+
+   !***  Wright (1997) constants from Table 1, last column
+
+   real (kind=RKIND), parameter ::      &
+      a0 = 7.057924e-4_RKIND,           &
+      a1 = 3.480336e-7_RKIND,           &
+      a2 = -1.112733e-7_RKIND
+
+   real (kind=RKIND), parameter ::      &
+      b0 = 5.790749e8_RKIND,            &
+      b1 = 3.516535e6_RKIND,            &
+      b2 = -4.002714e4_RKIND,           &
+      b3 = 2.084372e2_RKIND,            &
+      b4 = 5.944068e5_RKIND,            &
+      b5 = -9.643486e3_RKIND
+
+   real (kind=RKIND), parameter ::      &
+      c0 = 1.704853e5_RKIND,            &
+      c1 = 7.904722e2_RKIND,            &
+      c2 = -7.984422_RKIND,             &
+      c3 = 5.140652e-2_RKIND,           &
+      c4 = -2.302158e2_RKIND,           &
+      c5 = -3.079464_RKIND
```

Comment:
@cbegeman and @dengwirda, here are the definitions of the coefficients for the EOS.  Wouldn't hurt to sanity check them if you like, but don't feel you have to.  They are the last column in Table 1 in Wright (1997), the "Extended formula fit over the reduced range".  Above, I have limited T and S to this reduced range, except that I've let the T get extrapolated a bit below the minimum of -2 C because lower temperatures can be reached under ice shelves and I expect the EOS to be reasonably behaved with a little extrapolation.